### PR TITLE
feat(rome_js_formatter): trailing comma option #3305

### PR DESCRIPTION
**Pull request by denbezrukov**
## Summary

Add support for a [trailing_comma](https://prettier.io/docs/en/options.html#trailing-commas) option that can either be `es5` or` all` (default)

Close https://github.com/rome/tools/issues/3305

## Test Plan

Add new snapshots

`cargo test -p rome_js_formatter`

### CODE REVIEW

File: crates/rome_js_formatter/src/context/trailing_comma.rs [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+                } else {
+                    TrailingSeparator::Omit
+                }
+            }
+            TrailingCommaMode::ES5 => TrailingSeparator::Allowed,
+        }
+    }
+}
+
+impl Format<JsFormatContext> for TrailingCommaMode {
```
Comment (MichaReiser) [inline crates/rome_js_formatter/src/context/trailing_comma.rs:31]: Is the `Format` implementation used anywhere or do all calls use `trailing_separator`

File: crates/rome_js_formatter/src/context/trailing_comma.rs [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+    fn fmt(&self, f: &mut Formatter<JsFormatContext>) -> FormatResult<()> {
+        if matches!(self, TrailingCommaMode::ES5) || f.options().trailing_comma().is_all() {
+            write!(f, [if_group_breaks(&text(","))])?
+        }
+
+        Ok(())
+    }
+}
```
Comment (MichaReiser) [inline crates/rome_js_formatter/src/context/trailing_comma.rs:39]: I believe there are 3 more places where we insert trailing separators (array, call arguments, and I forgot xD). You can search for `if_group_breaks(&text(",")`

File: website/src/_includes/docs/project-config.md [hunk 3]
```diff
@@ -47,16 +47,16 @@ An array of Unix shell style patterns.
 
 #### `linter.rules.recommended`
 
-Enables the [recommended rules](/docs/lint/rules) for all the groups. 
+Enables the [recommended rules](/docs/lint/rules) for all the groups.
```
Comment (MichaReiser) [inline website/src/_includes/docs/project-config.md:50]: @ematipico is it necessary to split the documentation changes into its own PR that gets merged into the 10.0 release PR?

**Review by MichaReiser — APPROVED** [review]
Awesome work! There are three more places where trailing commas exist but think this is otherwise good to merge (search for `if_group_breaks(&text(",")`)

File: crates/rome_js_formatter/src/context/trailing_comma.rs (hunk 2 above)
Comment (denbezrukov) [reply crates/rome_js_formatter/src/context/trailing_comma.rs:39]: Oh,
You're right. I've missed them =(
done

File: crates/rome_js_formatter/tests/specs/js/module/array/trailing-comma/array_trailing_comma.js.snap [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+	dsadsadasdasdasdasdasdasdasd,
+] = [1, 2, 10];
+## Output 3
+-----
+Indent style: Tab
+Line width: 80
+Quote style: Double Quotes
+Quote properties: As needed
+Trailing comma: All
+-----
```
[the remark this answers is not in the document]
Comment (denbezrukov) [reply crates/rome_js_formatter/tests/specs/js/module/array/trailing-comma/array_trailing_comma.js.snap:66]: Done

File: crates/rome_js_formatter/src/context/trailing_comma.rs [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+
+#[derive(Default, Debug, Eq, PartialEq, Clone, Copy)]
+#[cfg_attr(
+    feature = "serde",
+    derive(serde::Serialize, serde::Deserialize, schemars::JsonSchema)
+)]
+pub enum TrailingComma {
```
[the remark this answers is not in the document]
Comment (denbezrukov) [reply crates/rome_js_formatter/src/context/trailing_comma.rs:46]: The difference between TrailingComma and TrailingCommaMode:
TrailingComma is a formatter option.
TrailingCommaMode is for using in format functions.

E.g. TrailingComma option can be ES5.

And TrailingCommaMode in arrow_function-expression can be All.

What do you think? We can't merge them because they are for different purpose.
maybe it's not the best naming

File: crates/rome_js_formatter/src/context/trailing_comma.rs (hunk 1 above)
Comment (denbezrukov) [reply crates/rome_js_formatter/src/context/trailing_comma.rs]: Some functions use TrailingComma as `Format`, some function which use `format_separated` for formatting use `trailing_separator` .

- https://github.com/rome/tools/pull/3430/files#diff-e5e1b0d9bea6928cd9f5230cfb0f3ad09ec281c3ad15cf6e02bbb98bd042c219R11-R19
- https://github.com/rome/tools/pull/3430/files#diff-8e7a5d166143d30b4e78f2bc2ede09676ffb63ee74a9a5013019861fe8ab88daR243

Comment (denbezrukov) [reply crates/rome_js_formatter/src/context/trailing_comma.rs]: Renamed

File: website/src/_includes/docs/project-config.md (hunk 3 above)
Comment (ematipico) [reply website/src/_includes/docs/project-config.md:50]: Ideally yes, we don't want to push changes that are not in sync with the latest release, unless we must. In this case we can just add reminder in #3319

Comment (denbezrukov) [reply website/src/_includes/docs/project-config.md:50]: Added.

File: crates/rome_js_formatter/tests/specs/js/module/array/trailing-comma/array_trailing_comma.js.snap [hunk 6]
```diff
@@ -0,0 +1,79 @@
...
+	longlonglonglongItem1longlonglonglongItem3,
+];
+
+// destructuring
+[
+	adsadasdasdasdasdasdasdasdasdasdas,
+	dsadsadasdasdasdasdasdasdasd,
+	dsadsadasdasdasdasdasdasdasd,
+] = [1, 2, 10];
+## Output 2
```
[the remark this answers is not in the document]
Comment (denbezrukov) [reply crates/rome_js_formatter/tests/specs/js/module/array/trailing-comma/array_trailing_comma.js.snap:39]: I will do so after this PR is merged.

File: website/src/_includes/docs/project-config.md (hunk 3 above)
Comment (MichaReiser) [reply website/src/_includes/docs/project-config.md:50]: Thanks. Can you revert these changes here. This is necessary to avoid that the documentation gets updated when merging the PR rather than when we release the new version.
